### PR TITLE
Create link formatting.adoc

### PR DESCRIPTION
**Pull request by John-Clifton-SonarSource**
Added in formatting guidance for links that appear in rules.

## Review

A dedicated reviewer checked the rule description successfully for:

- [x] logical errors and incorrect information
- [x] information gaps and missing content
- [x] text style and tone
- [x] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: docs/link_formatting.adoc [hunk 1]
```diff
@@ -0,0 +1,65 @@
+== Why is it important to define a standard for links?
+
+Rule text often wants to link to blog posts, documentation, and other rules. To provide a consistent experience across the Sonar solution, we want to define standard ways to write links in our rules. This document aims to do that.
+
+== How should you write links?
+
+=== Link formatting:
+* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:8]:
```suggestion

* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
```

File: docs/link_formatting.adoc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+* Links to blog posts / news articles: <source> - <title> e.g. Google Security Blog - Moving towards a more secure web
+* Links to stackoverflow answers (similar comments in a blog)e.g. StackOverflow - Answer by Stephen Cleary for Best way to handle null task inside async method?
+* Link to another Sonar rule: <rulenumber> - <title> e.g. S2755 - XML parsers should not be vulnerable to XXE attacks (note, linking to rules.sonarsource.com)
+
+The hyperlink to the article should be applied to just the document name, with the 'source' being left as plain text. The idea is that this makes it really easy for a user to understand the source before they click on anything.
+
+=== Additional things to consider when adding a link to a rule:
+* Is the article (particularly blog posts) likely to be around for at least 6 months?
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:16]:
```suggestion

* Is the article (particularly blog posts) likely to be around for at least 6 months?
```

File: docs/link_formatting.adoc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+* Do you trust this source as an 'expert'?
+* Link to en_US versions where there is a choice
+
+=== Short form names for sources we regularly use
+When web pages have massively long names like "Java™ Platform, Standard Edition 8 API Specification", we will provide short form names that we will use instead. Feel free to add new ones below:
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:21]:
```suggestion

When web pages have massively long names like "Java™ Platform, Standard Edition 8 API Specification", we will provide short form names that we will use instead. Feel free to add new ones below:
```

File: docs/link_formatting.adoc [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+Rule text often wants to link to blog posts, documentation, and other rules. To provide a consistent experience across the Sonar solution, we want to define standard ways to write links in our rules. This document aims to do that.
+
+== How should you write links?
+
+=== Link formatting:
+* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
+* Links to blog posts / news articles: <source> - <title> e.g. Google Security Blog - Moving towards a more secure web
+* Links to stackoverflow answers (similar comments in a blog)e.g. StackOverflow - Answer by Stephen Cleary for Best way to handle null task inside async method?
+* Link to another Sonar rule: <rulenumber> - <title> e.g. S2755 - XML parsers should not be vulnerable to XXE attacks (note, linking to rules.sonarsource.com)
```
Comment (necto) [inline docs/link_formatting.adoc:11]: What is the use case for explicit links to Sonar rules that is not covered by automatic wikification that converts rule ids to the links to these rules?

Comment (necto) [reply docs/link_formatting.adoc:11]: Should in this case the hyperlink be applied to the entire "<rulenumber> - <title>" or just "<title>"?

File: docs/link_formatting.adoc [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+The hyperlink to the article should be applied to just the document name, with the 'source' being left as plain text. The idea is that this makes it really easy for a user to understand the source before they click on anything.
+
+=== Additional things to consider when adding a link to a rule:
+* Is the article (particularly blog posts) likely to be around for at least 6 months?
+* Do you trust this source as an 'expert'?
+* Link to en_US versions where there is a choice
+
```
Comment (necto) [inline docs/link_formatting.adoc:19]: Should there be any consideration for the attribution of the source.
E.g., is there any issue with linking to branded sources, such as Microsoft, Google, Amazon blogs, articles and spin-offs?

File: docs/link_formatting.adoc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+== How should you write links?
+
+=== Link formatting:
+* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
+* Links to blog posts / news articles: <source> - <title> e.g. Google Security Blog - Moving towards a more secure web
+* Links to stackoverflow answers (similar comments in a blog)e.g. StackOverflow - Answer by Stephen Cleary for Best way to handle null task inside async method?
+* Link to another Sonar rule: <rulenumber> - <title> e.g. S2755 - XML parsers should not be vulnerable to XXE attacks (note, linking to rules.sonarsource.com)
+
+The hyperlink to the article should be applied to just the document name, with the 'source' being left as plain text. The idea is that this makes it really easy for a user to understand the source before they click on anything.
+
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:14]:
```suggestion
Here is how the above links should look like:

* SonarQube - https://docs.sonarqube.org/9.7/analyzing-source-code/scanners/sonarscanner-for-gradle/[SonarScanner for Gradle]
* Google Security Blog - https://security.googleblog.com/2016/09/moving-towards-more-secure-web.html[Moving towards a more secure web]
* StackOverflow - Answer by Stephen Cleary for https://stackoverflow.com/a/27551261[Best way to handle null task inside async method?]
* S2755 - https://rules.sonarsource.com/java/RSPEC-2755[XML parsers should not be vulnerable to XXE attacks]
```

File: docs/link_formatting.adoc (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply docs/link_formatting.adoc:8]: I would add an explanation that this does not apply to inline links used in sentences. This format applies to links in the Resources section.

File: docs/link_formatting.adoc [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+=== Additional things to consider when adding a link to a rule:
+* Is the article (particularly blog posts) likely to be around for at least 6 months?
+* Do you trust this source as an 'expert'?
+* Link to en_US versions where there is a choice
+
+=== Short form names for sources we regularly use
+When web pages have massively long names like "Java™ Platform, Standard Edition 8 API Specification", we will provide short form names that we will use instead. Feel free to add new ones below:
+
+==== Short form names to use:
+* Android Documentation - https://developer.android.com/docs
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:24]:
```suggestion

* Android Documentation - https://developer.android.com/docs
```

File: docs/link_formatting.adoc [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+* Sonar - https://www.sonarsource.com/
+* Sonar Blog - https://www.sonarsource.com/blog/
+* Stack Overflow - https://stackoverflow.com/
+* Symfony - https://symfony.com/doc/current/index.html
+* Test NG Documentation - https://testng.org/doc/documentation-main.html
+* W3Schools - https://www.w3schools.com/
+* WCAG  - https://www.w3.org/WAI/standards-guidelines/wcag/
+* Wikipedia - https://en.wikipedia.org
+
+
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:65]:
```suggestion
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: docs/link_formatting.adoc (hunk 5 above)
Comment (John-Clifton-SonarSource) [reply docs/link_formatting.adoc:20]: We don't currently have any guidance on this. If that proves to be necessary we will update in the future.

File: docs/link_formatting.adoc (hunk 4 above)
Comment (John-Clifton-SonarSource) [reply docs/link_formatting.adoc:12]: Good thinking on how we do the hyperlink for rules. I have updated.

I want to provide the opportunity for rules to link to other rules explicitly in the Resources section. This signals to users that they might want to read about a particular related rule as a way to grow their skills when they aren't worrying about getting the current issue fixed.

File: docs/link_formatting.adoc [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+* PEP - https://peps.python.org/
+* PHP Documentation - https://www.php.net/docs.php
+* PHP Tutorials - https://www.phptutorial.net/
+* PEP 8 – Style Guide for Python Code - https://peps.python.org/pep-0008/
+* Python Documentation - https://docs.python.org/
+* React Documentation - https://reactjs.org/
+* Rhino Security Labs - https://rhinosecuritylabs.com/
+* SANS  -https://www.sans.org/
+* SAP Documentation - http://help.sap.com/abapdocu_702/en/abenabap.htm
+* SonarQube Documentation -  https://docs.sonarqube.org/latest/
```
Comment (sebastien-marichal) [inline docs/link_formatting.adoc:66]: This short name does not match the example [here](https://github.com/SonarSource/rspec/pull/2324/files#diff-47a3cede129c305289caab3167eecb0d7aea987ab2fd51eef2967ba347a76192R19):
```
* SonarQube - https://docs.sonarqube.org/9.7/analyzing-source-code/scanners/sonarscanner-for-gradle/[SonarScanner for Gradle]
```

File: docs/link_formatting.adoc [hunk 10]
```diff
@@ -0,0 +1,75 @@
+== Why is it important to define a standard for links?
+
+We often want to provide links in the 'Resources' section of a rule that point to blog posts, documentation, and other rules. To provide a consistent experience across the Sonar solution, we want to define standard ways to write these links in our rules. This document aims to do that. This guidance does not apply to links that appear within continuous text elsewhere in a rule.
+
+== How should you write links?
+
+=== Link formatting:
+
+* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:9]:
```suggestion
* Links to documentation: <source> - <title> e.g. SonarQube - SonarScanner for Gradle
* Links to framework documentation <source> - <member name and kind> e.g. Microsoft Learn - Object.ToString Method
```

File: docs/link_formatting.adoc [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+* Links to blog posts / news articles: <source> - <title> e.g. Google Security Blog - Moving towards a more secure web
+* Links to stackoverflow answers (similar comments in a blog)e.g. StackOverflow - Answer by Stephen Cleary for Best way to handle null task inside async method?
+* Link to another Sonar rule: <rulenumber> - <title> e.g. S2755 - XML parsers should not be vulnerable to XXE attacks (note, linking to rules.sonarsource.com)
+
+The hyperlink to anything apart from other Sonar rules should be applied to just the document name, with the 'source' being left as plain text. The idea is that this makes it really easy for a user to understand the source before they click on anything.
+For Sonar rules, the whole entry (<rulenumber> - <title>) should be a hyperlink.
+
+Here is how the above links should look like:
+
+* SonarQube - https://docs.sonarqube.org/9.7/analyzing-source-code/scanners/sonarscanner-for-gradle/[SonarScanner for Gradle]
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:19]:
```suggestion
* SonarQube - https://docs.sonarqube.org/9.7/analyzing-source-code/scanners/sonarscanner-for-gradle/[SonarScanner for Gradle]
* Microsoft Learn - `Object.ToString` Method[https://learn.microsoft.com/en-us/dotnet/api/system.object.tostring]
```

File: docs/link_formatting.adoc [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+* Google Security Blog - https://security.googleblog.com/2016/09/moving-towards-more-secure-web.html[Moving towards a more secure web]
+* StackOverflow - Answer by Stephen Cleary for https://stackoverflow.com/a/27551261[Best way to handle null task inside async method?]
+* S2755 - https://rules.sonarsource.com/java/RSPEC-2755[XML parsers should not be vulnerable to XXE attacks]
+=== Additional things to consider when adding a link to a rule:
```
Comment (martin-strecker-sonarsource) [inline docs/link_formatting.adoc:23]:
```suggestion

=== Additional things to consider when adding a link to a rule:
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
Looks good to me. I think it would be valuable to also add one example for links to framework APIs (I suggested Object.ToString) as it is one of the most common links we use.

File: docs/link_formatting.adoc (hunk 9 above)
Comment (John-Clifton-SonarSource) [reply docs/link_formatting.adoc:67]: Nice catch, thank you. I have updated that and another instance of the same thing.

File: docs/link_formatting.adoc (hunk 10 above)
Comment (John-Clifton-SonarSource) [reply docs/link_formatting.adoc]: Manually merged the change.

File: docs/link_formatting.adoc (hunk 11 above)
Comment (John-Clifton-SonarSource) [reply docs/link_formatting.adoc]: Manually merged the change.

**Pull request merged**